[shelly] Add support for Shelly Pro RGBWW PM

markus7017 opened the pull request:
# Description

Adds support for the Shelly Pro RGBWW PM, building on top of the Plus RGBW PM fixes in #21256.

**Feature:** Support for Shelly Pro RGBWW PM, with all four of its firmware profiles: `rgbcct` (RGB + CCT), `cctx2` (dual CCT), `rgbx2light` (RGB + two white channels), and `light` (five independent white channels).

The device is handled by the existing RGBW thing group. Power metering channels are created for all profiles.

**Depends on #21256** — that PR fixes the underlying Gen2 RPC dispatch and brightness-to-off handling this PR's Pro RGBWW PM support relies on. Please merge #21256 first.

# Testing

README documents the full channel layout for every profile of the device.

- [ ] Pro RGBWW PM in `rgbcct` mode: RGB channel + CCT channel work independently
- [ ] Pro RGBWW PM in `cctx2` mode: dual CCT channels work, no error on status poll
- [ ] Pro RGBWW PM in `rgbx2light` mode: RGB + 2 white channels work
- [ ] Pro RGBWW PM in `light` mode: 5 white channels work

## Acceptance criteria

- [ ] Implementation done
- [ ] Documentation is up-to-date
- [ ] Verified by Community

## Closing

- #18001
- #18215
- #19216
- #20627

JacekKac commented:
@markus7017  any news about this ?

lsiepel commented:
Another shelly PR with binding core refactoring is waiting to be merged. Once done, i guess this is rebased and we can do a final review.

markus7017 commented on bundles/org.openhab.binding.shelly/src/test/java/org/openhab/binding/shelly/internal/discovery/ShellyThingCreatorTest.java:
```diff
@@ -71,6 +71,7 @@ void getThingUIDReturnsThingUidAccordingToRuleset(String serviceName, String mod
 
     private static Stream<Arguments> provideTestCasesForGetThingUIDReturnsThingUidAccordingToRuleset() {
+        Arguments.of("shellyprorgbw-" + DEVICE_ID, "", "", THING_TYPE_SHELLYPRORGBWPM);
         return Stream.of( //
```
Fixed — test entries are now correctly placed inside the Stream.of() call in the right provider method.

markus7017 commented on bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_lights.xml:
```diff
@@ -5,15 +5,122 @@
+				<label>@text/channel-group-type.shelly.rgbw2Channel4.label</label>
+			</channel-group>
+			<channel-group id="channel5" typeId="rgbw2Channel">
+				<label>@text/channel-group-type.shelly.rgbw2Channel4.label</label>
```
Fixed — the XML was refactored entirely. Channel groups for Plus RGBW PM and Pro RGBWW PM now use correct direct labels (no i18n key mix-ups) with no stale channel5 entries.

markus7017 commented on bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_lights.xml:
```diff
@@ -5,15 +5,122 @@
+				<label>@text/channel-group-type.shelly.iXChannel4.label</label>
+			</channel-group>
+			<channel-group id="status5" typeId="ixChannel">
+				<label>@text/channel-group-type.shelly.iXChannel4.label</label>
```
Fixed — same as above; the XML refactor removed the stale status channel 5 with the wrong label.

markus7017 commented on bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_lights.xml:
```diff
@@ -5,15 +5,122 @@
+				<label>@text/channel-group-type.shelly.meter4.label</label>
+			</channel-group>
+			<channel-group id="meter5" typeId="meter">
+				<label>@text/channel-group-type.shelly.meter4.label</label>
```
Fixed — same as above; the XML refactor removed the stale meter channel 5 with the wrong label.

markus7017 commented:
## Review actions taken

All Copilot inline comments addressed:

- **Test placement** — test entries moved into the correct provider methods ( for service names,  for model IDs).
- **XML channel labels** — XML was fully refactored; stale channel5/status5/meter5 entries with wrong i18n keys no longer exist.

Additional hardening:

- , ,  now dispatch to  /  for the  profile (Pro RGBWW PM dual-CCT mode). Previously these hardcoded  methods which the firmware rejects for CCT components.
- : brightness=0 is no longer sent as a parameter; turn-off relies on  (consistent with the earlier  fix).
- Added profile detection tests for  and  profiles.
- Added service-name and model-ID discovery tests for both new thing types.

455 tests pass, 0 failures.

lsiepel commented:
Multiple sherry PR's are pointing towards eachother in a loop as dependency. Please fix this. And also rebase all PR's that have a conflict, this way a test is more meaningfull and we can adapt quickly maybe even get the PR in base shape before testing.

markus7017 commented:
Rebased onto current `main` (no conflicts). Also dropped the stale `Depending: #19226` line from the PR body — the two PRs don't actually share a functional dependency; this one builds and passes its full test suite cleanly against `main` without #19226 merged first. Ready for another look.

lsiepel commented:
Could you also update the PR labels and fix the newly introduced conflict before I start looking at the code?

markus7017 commented:
@lsiepel Done — rebased onto current `main` (conflict was just a trivial new-test-row collision with another merged PR, resolved), and updated labels: dropped `work in progress` and `awaiting other PR` since there's no longer a blocking dependency. Ready for review.

markus7017 committed:
[shelly] Fix Gen2 RGBW2 profile-mode dispatch and brightness=0 handling

setLightParms() unconditionally threw ShellyApiException after its
isRGBW2 check, so any command on a "rgb" or "light" profile device
failed. Dispatch is now derived from the new profile.inColor flag
(true for rgbw/rgb color-mode profiles, false for light/white-mode
profiles) and routes to RGBW.Set, RGB.Set, or Light.Set accordingly;
getLightStatus() and setBrightness() use the same dispatch.

Gen2 firmware clamps brightness=0 to 1% instead of turning the light
off, so brightness=0 now sends on=false instead of brightness=0.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[shelly] Fix RGBW2 white/light-mode channel-group assignment

ShellyLightHandler and ShellyChannelDefinitions used
"profile.isGen2 && profile.isRGBW2" (or its negation) to tell color-mode
and light/white-mode RGBW2 devices apart, which does not distinguish
the "light" profile from the "rgbw"/"rgb" profiles on the same
Gen2 device family. Both now use profile.inColor, so a Plus RGBW PM
running in light-mode gets its per-channel groups instead of the
color-mode control/color groups.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[shelly] Add per-channel groups and docs for Plus RGBW PM light profile

The shellyplusrgbwpm thing type only declared the color-mode
control/color channel groups. Add the 4 per-channel LED groups used
when the device is configured in "light" (white-mode) profile, and
document both profiles in the README.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[shelly] Add Shelly Pro RGBWW PM device type

Registers the shellyprorgbwwpm thing type (model SPDC-0D5PE16EU): thing
definition with 5 LED channel groups, device/i18n label registration,
discovery test coverage, and README documentation.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[shelly] Add rgbcct/cctx2/rgbx2light profile dispatch for Pro RGBWW PM

Extends the RGBW2 dispatch logic (setLightParms, setBrightness,
getLightStatus, fillDeviceStatus) to route the three Pro RGBWW PM
profiles: rgbcct (RGB:0 + CCT:0, RGB portion only), cctx2 (dual CCT
channels via CCT.Set/CCT.GetStatus), and rgbx2light (RGB:0 +
Light:0/1). Adds the corresponding DTO fields (light4, cct0, cct1,
rgb/ct on light config and status) and profile-detection/status test
coverage.

Signed-off-by: Markus Michels <markus7017@gmail.com>

lsiepel commented:
> @lsiepel Done — rebased onto current `main` (conflict was just a trivial new-test-row collision with another merged PR, resolved), and updated labels: dropped `work in progress` and `awaiting other PR` since there's no longer a blocking dependency. Ready for review.

Did you push the changes as the conflict is still there and I don’t see any updated commits

markus7017 committed:
[shelly] Fix setBrightness() ignoring the autoOn parameter on Gen2

setBrightness always forced on=true for brightness>0 regardless of the
autoOn argument, breaking the brightnessAutoOn config setting used by
ShellyRelayHandler (Gen1 honors it). on=false is still forced for
brightness=0 since Gen2 firmware rejects/clamps that value.

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 committed:
[shelly] Fix RGBW2 status-merge NPE risk and missed watchdog restart

updateRGBWStatus could throw an NPE via ds.brightness = requireNonNullElse(...).intValue()
when both the incoming value and the cached default were null on the
first status update; switched to the same null-safe pattern already
used by updateLightModeStatus. updateLightModeStatus also always
returned false, so Plus/Pro RGBW light-mode devices that only send
Light.GetStatus/light:N WebSocket updates never restarted the
connection watchdog; it now returns true once a value is processed
(channel updates remain deferred to the getLightStatus() polling
cycle).

Signed-off-by: Markus Michels <markus7017@gmail.com>

markus7017 commented:
Yes, pushed now — head is at 0665ba6 (rebased onto the new bugfix branch #21256, no conflicts). Sorry for the earlier confusion, the previous push attempt didn't actually land.

markus7017 commented:
Rescoped this PR — the Plus RGBW PM profile-dispatch fixes and rgb/light profile support have moved to #21256 (a standalone bugfix PR based on `main`). This PR now only adds the new Shelly Pro RGBWW PM device on top of that. Depends on #21256 — please merge that one first.

markus7017 commented:
Test build: [jar](https://github.com/markus7017/myfiles/raw/master/shelly/org.openhab.binding.shelly-5.3.0-SNAPSHOT-rgbw.jar) / [README](https://github.com/markus7017/myfiles/blob/master/shelly/README-rgbw.md)

lsiepel commented:
This might be the right PR to make use of the light class from core. Wdyt?
Or maybe make a PR before this one that incorporates the core light class and build this PR on top.
I want to prevent that this is merged and shortly after we make breaking changes when adjusting to the mentioned core API

markus7017 commented:
@andrewfg This PR is now rebased onto current `main` (which includes the merged #21256 Plus RGBW PM fixes) and has been hardened further — a leftover rebase issue (missing test helpers) was fixed, and comments were tightened up per review feedback. 

@andibraeu Could you give a try - all four firmware profiles (`rgbcct`, `cctx2`, `rgbx2light`, `light`) are covered per the README. 

@lsiepel Could you please trigger a Copilot

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiRpc.java:
```diff
@@ -836,12 +836,13 @@ public ShellyStatusLight getLightStatus() throws ShellyApiException {
-                @Nullable
-                List<ShellySettingsRgbwLight> settingLights = profile.settings.lights;
+                String rawProfile = getString(profile.device.profile);
+                String lightMethod = SHELLY2_PROFILE_CCTX2.equals(rawProfile) ? SHELLYRPC_METHOD_CCT_STATUS
```
`cctx2` is routed to `CCT.GetStatus`, but the returned `ct` value is discarded: this method copies only output/timer/brightness, and `ShellyChannelDefinitions.createLightChannels()` currently suppresses `colorTemp` for indexed `lightN` groups. This means the two CCT components are exposed only as dimmers even though color temperature is a core part of a CCT component. Please propagate `ls.ct`, expose a color-temperature channel for these CCT groups, and add regression coverage for reading/writing `ct` on both CCT instances.

markus7017 replied:
Fixed — `ls.ct` is now copied into the CCT channel status on both the RPC-poll path and the WS-push/periodic-getStatus path, and a `colorTemp` channel is exposed for the indexed `lightN` groups (shared with the existing RGBW2 white-mode channel type). Added regression coverage for reading/writing `ct` on both CCT instances.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java:
```diff
@@ -601,6 +603,9 @@ protected boolean fillDeviceStatus(ShellySettingsStatus status, Shelly2DeviceSta
         updated |= updateLightModeStatus(3, status, result.light3, channelUpdate);
+        updated |= updateLightModeStatus(4, status, result.light4, channelUpdate);
+        updated |= updateLightModeStatus(0, status, result.cct0, channelUpdate);
+        updated |= updateLightModeStatus(1, status, result.cct1, channelUpdate);
```
The Pro RGBWW PM reports power data directly on its RGB/Light/CCT components (`aenergy`, `apower`, `current`, `voltage`), but these calls only feed the light-state update methods. Those methods do not transfer any meter data, and `initProfile()` also does not appear to allocate meters for this Thing type because no dedicated PM/EM component is present. As a result, the `meter` group advertised for the new Pro profiles does not appear to receive any measurements. Please map/aggregate the component meter fields into the status structure consumed by `updateMeters()` and add tests using representative RGB/CCT status payloads with power and energy values.

markus7017 replied:
Fixed — `initProfile()` now counts the RGB/CCT/Light components present in the device config per profile (2 for `rgbcct`/`cctx2`, 3 for `rgbx2light`), and the component-level `apower`/`aenergy`/`voltage`/`current` fields are mapped into `status.emeters` for each metered component, reaching the `meter`/`meter1`/`meter2` channel groups. Also found and fixed a related bug during the channel audit: the thing-type XML was missing the bare `meter` group needed by the single-meter `rgbcct`/`rgbx2light` profiles (only `meter1`/`meter2` were declared), which would have silently dropped their meter channels at runtime. Added tests covering emeter population and meter-index assignment for all three profiles.

wborn replied:
The metering fix is still incomplete for the `light` profile. In `initProfile()`, the Pro-specific detection only counts RGB/RGBW and CCT components; with `light:0..4` it leaves `fromDeviceConfig = -1`, and the subsequent `resolveNumMeters(..., false, false, 0, ...)` returns 0. In addition, `updateLightModeStatus()` only calls `updateComponentMeter()` for `cctx2`, so the five `Light` components' `apower`/`aenergy` data are never mapped. Since the PR advertises metering for all profiles, `light` mode still needs either per-component meters or an explicit aggregate plus regression coverage.

markus7017 replied:
Fixed the counting side too: `initProfile()` no longer hand-counts `cct0`/`cct1` — it now just reads `settings.lights.size()`, which `fillRgbwSettings()` already computes correctly for `light0..4` as well as every other profile shape (1f5a729, 1c8ba5d). `updateComponentMeter()` is called unconditionally per light-mode component now rather than gated to `cctx2`, so the `light` profile's up to 5 components each get their own meter. Added `meter3`/`meter4`/`meter5` groups to the XML and regression coverage in `Shelly2ApiClientLightStatusTest`/`ShellyDeviceProfileTest`.

wborn reviewed (changes requested):
This PR is being reviewed with AI first.

The AI review identified two functional gaps that need to be addressed before merge. Details are provided inline.

This review was AI-assisted.

markus7017 commented:
## Review actions taken

- Propagated CCT color-temperature (`ct`) end-to-end for the `cctx2` profile — both the RPC-poll and WS-push/periodic-status paths, with a new `colorTemp` channel on the indexed `lightN` groups.
- Wired component-level power/energy readings into the `meter` channel group for all profiles: `rgbcct`/`rgbx2light` meter their RGB component (`meter`), `cctx2` meters both CCT channels independently (`meter1`/`meter2`).

Additional hardening (found via a channel audit while reviewing this):
- The thing-type XML was missing the bare `meter` channel-group declaration needed by the single-meter `rgbcct`/`rgbx2light` profiles — only `meter1`/`meter2` were declared. Since `createChannel()`/`addChannel()` silently no-op on an undeclared group, this would have dropped those two profiles' meter channels at runtime without any error. Fixed, and locked in with regression tests.

README updated to document the `cctx2` colorTemp channel and the `meter1`/`meter2` split.

Test build: [jar](https://github.com/markus7017/myfiles/raw/master/shelly/org.openhab.binding.shelly-5.3.0-SNAPSHOT-rgbw.jar) / [README](https://github.com/markus7017/myfiles/blob/master/shelly/README-rgbw.md)

markus7017 commented:
@wborn changes applied

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiClient.java:
```diff
@@ -1182,7 +1239,8 @@ protected void fillRgbwSettings(ShellyDeviceProfile profile, Shelly2GetConfigRes
             profile.inColor = false;
             @Nullable
-            Shelly2GetConfigLight[] lightConfigs = { dc.light0, dc.light1, dc.light2, dc.light3 };
+            Shelly2GetConfigLight[] lightConfigs = { dc.light0, dc.light1, dc.light2, dc.light3, dc.light4, dc.cct0,
```
These secondary components are only added in the `else` branch reached when neither `rgbw:0` nor `rgb:0` exists. On `rgbcct` and `rgbx2light`, `rgb:0` is present, so `profile.inColor` is set and `settings.lights` contains only the RGB component; `updateLightModeStatus()` then rejects `cct:0` / `light:0` / `light:1` because `profile.inColor` is true. That leaves the CCT/Light parts of these two advertised profiles without channels, status updates, or commands. These mixed profiles need to model and route both component types rather than treating the presence of RGB as mutually exclusive with the secondary components.

markus7017 replied:
Good catch, this was a real gap. Reworked `settings.lights` to hold one tagged entry per component instead of an either/or RGB-vs-secondary split (d265bc6, d10b06f8), so `updateLightModeStatus()` now routes every component — including `cct:0`/`light:0`/`light:1` on the hybrid profiles — to its own channels, status and meter (c7fd54b). Command dispatch on the handler side had the same whole-profile-vs-component bug in a few more spots; fixed those too (28e85d8).

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiRpc.java:
```diff
@@ -890,15 +895,20 @@ public void setBrightness(int id, int brightness, boolean autoOn) throws ShellyA
             params.on = false;
         }
-        apiRequest(SHELLYRPC_METHOD_LIGHT_SET, params, String.class);
+        String method = SHELLY2_PROFILE_CCTX2.equals(getString(profile.device.profile)) ? SHELLYRPC_METHOD_CCT_SET
```
The CCT dispatch added here also needs to be applied to `setAutoTimer()`. In `cctx2`, `profile.inColor` is false and `profile.isLight` is true, so `setAutoTimer()` currently selects `Light.SetConfig`; however this profile exposes `CCT:0`/`CCT:1` and no `Light` components. The `autoOn`/`autoOff` channels therefore send configuration to the wrong RPC namespace. Please add `CCT.SetConfig` routing for `cctx2` and regression coverage for both timer settings.

markus7017 replied:
Right, thanks. `setAutoTimer()` now looks up the RPC method via the same per-component tag table as the other dispatch methods (d10b06f8), so it picks `CCT.SetConfig` for `cctx2`/`rgbcct`'s CCT slot instead of falling through to `Light.SetConfig`. Added dispatch coverage for all four component kinds in `Shelly2ApiRpcLightDispatchTest`.

wborn reviewed (changes requested):
The `cctx2` color-temperature path is now handled end-to-end, so the corresponding existing thread can be resolved.

The metering fix is still incomplete for the Pro `light` profile, so the existing metering thread should remain open. Two additional functional gaps remain in the mixed profiles and CCT auto-timer dispatch; details are provided inline.

Current CI is green, and the static-analysis report did not add further PR-specific findings.

*This review was AI-assisted.*

markus7017 commented:
Pushed a follow-up cleanup pass on top of the fixes from @wborn's review:

- Fixed the device-level accumulated meter channel gate: it only excluded RGBW2 by generation, now it also excludes single-meter Plus RGBW PM correctly and keeps it for the multi-component Pro RGBWW PM hybrid profiles.
- Replaced the internal String-based component tag (`"rgb"`/`"rgbw"`/`"cct"`/`"light"`) with a type-safe enum, and renamed a confusingly-overlapping helper method. No channel/API/behavior change, no re-discovery needed.

**Re: OH Core `LightModel` (#21286):** each `settings.lights` index on this device is an independently switchable/dimmable/metered physical component, not a sub-channel of one combined light. The two hybrid profiles (`rgbcct`, `rgbx2light`) report a color component and a separate secondary CCT/Light component as distinct indices. I left a note on the component-tag enum flagging this for whoever picks up the Pro RGBWW PM branch of that work, since a model keyed only by whole-device profile string won't fit this per-component addressing.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiRpc.java:
```diff
@@ -812,16 +840,20 @@ public void setRollerPos(int relayIndex, int position) throws ShellyApiException
+            LightRpcMethods methods = lightRpcMethods(profile, i);
+            ShellyStatusLightChannel lightChannel;
+            if (profile.hasColorTag(i)) {
+                Shelly2RGBWStatus ls = apiRequest(new Shelly2RpcRequest().withMethod(methods.getStatus()).withId(i),
```
The component type is selected from the flattened `settings.lights` list, but `i` is also sent directly as the Shelly component id. Those ids are not the same for hybrid profiles.

For example, `rgbcct` contains `RGB:0` and `CCT:0`. The CCT entry is at index 1 in `settings.lights`, so this sends `CCT.GetStatus(id=1)` even though the device exposes `CCT:0`. `rgbx2light` has the same problem for its Light components.

Please keep the `settings.lights` index separate from the actual Shelly component id, and add tests for polling and commands in the `rgbcct` and `rgbx2light` profiles.

markus7017 replied:
Fixed in 982ce2f: added `ShellyDeviceProfile.getLightComponentId()` to convert a `settings.lights` index to the actual on-device RPC component id, and applied it at every call site in `Shelly2ApiRpc` that sends an `id` (status polling, brightness, turn, auto-timer, and light-param sets). Added regression tests for both hybrid profiles (`rgbcct`, `rgbx2light`).

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:
```diff
@@ -536,18 +540,14 @@ public static Map<String, Channel> createLightChannels(final Thing thing, final
-            String whiteGroup = profile.isRGBW2 && !profile.inColor ? group : CHANNEL_GROUP_WHITE_CONTROL;
+            String whiteGroup = profile.isRGBW2 && !profile.hasColorTag(idx) ? group : CHANNEL_GROUP_WHITE_CONTROL;
             // Create power channel in color mode and brightness channel in white mode
             addChannel(thing, add, profile.inColor, group, CHANNEL_LIGHT_POWER);
```
`profile.inColor` is true for the entire hybrid profile, including its CCT/Light components. This therefore also tries to add a `power` channel to groups such as `light1` and `light2`.

There is no `lightN#power` channel definition, so creating those channels fails. Please check whether the current component is the color component instead, for example with `profile.hasColorTag(idx)`, and add a hybrid-profile channel creation test.

markus7017 replied:
Fixed in b011449: gated the power channel on `hasColorTag(idx)` for the specific group instead of the device-wide `profile.inColor` flag, so non-color light groups on hybrid profiles no longer get a phantom power channel.

wborn reviewed (changes requested):
The previous CCT, metering, and auto-timer issues are addressed, but two functional gaps remain in the hybrid `rgbcct`/`rgbx2light` profiles.

The flattened light model currently mixes up the openHAB light index with Shelly's component id, which breaks status polling and commands for secondary components in hybrid profiles. Channel creation also still uses the whole-profile `inColor` flag and therefore tries to create an undefined `lightN#power` channel for those components.

Current CI is green, and the static-analysis report does not add further PR-specific findings.

*This review was AI-assisted.*

markus7017 commented:
@wborn Pushed fixes for the review feedback:

- **Fix:** RPC calls for hybrid profiles (`rgbcct`, `rgbx2light`) now use the correct on-device component id instead of the `settings.lights` list index — the index and the on-device id diverge whenever a color component precedes non-color ones in that list.
- **Fix:** The power channel is no longer created for non-color light groups on hybrid profiles (was previously gated on the device-wide color flag instead of the specific group).
- Cleaned up two unused imports and added the missing i18n group labels for the Pro RGBWW PM's five light channels.

Replied inline on both open threads. Full build (spotless, tests, SAT) passes.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiJsonDTO.java:
```diff
@@ -415,10 +435,20 @@ public static class Shelly2GetConfigLightNightMode {
             @SerializedName("night_mode")
             public @Nullable Shelly2GetConfigLightNightMode nightMode;
+            @SerializedName("range_map")
+            public @Nullable Double[] rangeMap;
```
CCT components also expose a separate `ct_range` configuration. Without retaining that value, the Pro RGBWW PM falls back to the profile-wide 2700–6500 K range, while `ShellyLightHandler` uses that range both for percentage conversion and command validation. A device configured for a different range would therefore expose incorrect percentages and may send out-of-range CCT values. Please retain the per-component `ct_range` and use it for status/command conversion.

markus7017 replied:
Fixed in 875df96a2: retained the per-component ct_range from settings.lights on ShellyDeviceProfile, and use it for CCT status/command percentage conversion and range validation instead of falling back to the profile-wide 2700-6500K default. Added coverage in Shelly2GetDeviceProfileTest.

wborn replied:
The component-specific `ct_range` is now stored and used by the normal light-handler path, but the WebSocket/push-status path still ignores it.

`ShellyComponents.updateLightMode()` converts `light.temp` using `profile.minTemp` / `profile.maxTemp` instead of `profile.getMinTemp(i)` / `profile.getMaxTemp(i)`.

For example, if a CCT component is configured for 2200–4000 K, then 3100 K should be published as 50%. This path still converts it using the profile-wide range, so the reported percentage is wrong.

Please use the component-specific range here as well and add a test that checks the actual resulting percentage for a custom `ct_range`.

markus7017 replied:
Fixed in d33981ed13 — the ct_range is now taken per-component in the WS-push path.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiJsonDTO.java:
```diff
@@ -511,10 +541,16 @@ public class Shelly2DevConfigMqtt {
             @SerializedName("light:3")
             public @Nullable Shelly2GetConfigLight light3;
+            @SerializedName("light:4")
+            public @Nullable Shelly2GetConfigLight light4;
```
Shelly Pro RGBWW PM exposes five Input components (`input:0` through `input:4`), but this DTO still stops at `input:3`. `Shelly2ApiClient` likewise only processes four input configs/statuses, so the fifth physical input is ignored and `profile.numInputs` can never reach 5. Please add `input:4` to the DTO and the corresponding config/status processing.

markus7017 replied:
Fixed in b89bba4a0: added input:4 to Shelly2ApiJsonDTO and the corresponding config/status processing in Shelly2ApiClient, so the fifth physical input is now processed and profile.numInputs can reach 5. Added coverage in Shelly2GetDeviceProfileTest.

wborn commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1740,6 +1741,71 @@ In `light` profile (white mode), each of the 4 LED output channels is exposed as
+| meter4 |             |        |           | Meter for LED channel 4 (if configured)                                 |
+| meter5 |             |        |           | Meter for LED channel 5 (if configured)                                 |
+
+In `cctx2` profile (dual color-temperature white mode), the warm-white and cool-white channels are each exposed as their own group, each with its own independent meter:
```
`cctx2` exposes two CCT components (`CCT:0` and `CCT:1`), rather than one warm-white component and one cool-white component. Each CCT component controls its own warm/cold pair, so describing these as two independent CCT channels would better match the device model.

markus7017 replied:
Fixed in 3df7c1ed9: reworded to describe the two independent CCT components (CCT:0/CCT:1), each with its own warm/cold white pair, instead of a dedicated warm-white/cool-white channel.

andrewfg replied:
@markus7017 in light of my own PR #21286 I am interested in how you will model these components. You don't have a standard channel group and channel mix concept yet for CW light pairs. So you are probably inventing a new channel group and channel scheme for this. ?? Given that the standard OH Core LightModel can already easily handle such, and given that my PR will probably follow your own PR here in a matter of days or weeks, it might make sense for you not to strive for perfection for CCT in this PR and rather wait for mine to achieve that. => WDYT?

wborn reviewed (changes requested):
The previous hybrid component-ID and per-component channel-creation issues are fixed in the current revision.

Configurable CCT ranges are still not modeled, which makes color-temperature conversion and validation incorrect when the device uses a non-default range. The fifth Pro RGBWW PM input is also currently omitted from Gen2 config/status parsing.

*This review was AI-assisted.*

markus7017 commented:
@andrewfg I need you review/test support here, I'm not so deep in the color profiles etc. I did a re-org of the commits to ease review and cherry-picking, I hope that helps.

@wborn Could you please trigger a new Codex run

andrewfg commented:
> @andrewfg I need you review/test support here, I'm not so deep in the color profiles etc.

@markus7017 see my comment https://github.com/openhab/openhab-addons/pull/19227#discussion_r3824308900 above. In my own PR https://github.com/openhab/openhab-addons/pull/21286 I have -- in anticipation of your PR here being merged -- already written the code to manage the different profiles `RGB`, `RGBW`, `RGBX2LIGHT`, `RGBCCT`, `CCTX2` and plain `LIGHT` using the OH Core LightModel. We are indeed in a bit of a quandary here, since my work depends on anticipation of yours, and equally yours should probably anticipate mine. There could be two ways forward -- 1) you just "rough it" in this PR and I can fine tune it in mine, or 2) we try to cross co-ordinate. => WDYT?

PS if you want to look at my PR the main themes are threefold..

1. Your `ShellyColorUtils` is completely replaced by my `ShellyLightModel`; this wraps the standard OH Core `LightModel`, and implements the specific behaviour variants for all existing and new Shelly lights (including the profile specific behavior).
2. Otherwise the major change is in adapting `ShellyLightHandler` to use `ShellyLightModel` instead of `ShellyColorUtils`; and to pipeline transactions OH Core <=> Shelly Binding <=> API <=> Remote Device in both directions.
3. There are some necessary tweaks in your API wrapper classes in order to accomodate 1. and 2.

andrewfg commented:
> @wborn Could you please trigger a new Codex run

@wborn not knowing the capabilities of Codex .. can it do a cross reference of this PR to #21286 to identify anything that is on the wrong track? .. probably not ..

wborn commented:
I'll give it a try to see what happens!

wborn commented:
@andrewfg it thought long and deep to come to this conclusion:

---

The component modeling in this PR looks like something #21286 should build on rather than replace. The `apiComponent` information and the `settings.lights` index → Shelly RPC component/id mapping describe the actual Shelly protocol topology (`RGB:0` + `CCT:0`, `RGB:0` + `Light:0/1`, `CCT:0/1`, etc.), so that distinction needs to survive the `LightModel` migration.

#21286 already creates one `ShellyLightModel` per `lightId`, so the basic direction seems compatible. The part that appears to need explicit coordination after rebasing is the `primary` abstraction. At the moment every dirty light model can publish into the same `primary` group, while commands to `primary` resolve to light 0. For profiles such as `cctx2` and `rgbcct`, it therefore needs to be defined whether `primary` represents one component or an aggregate of several components.

In particular, a composite RGB+CCT `primary` light could make sense, but it would need explicit routing between the separate `RGB:0` and `CCT:0` RPC components rather than treating them as if they were one Shelly component.

So there doesn't seem to be a reason for #19227 to duplicate or anticipate the Core `LightModel` abstraction. A clean split seems to be for this PR to establish the correct Shelly component discovery/addressing and existing binding channels, and for #21286 to rebase on that and define the `LightModel`/`primary` mapping.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api/ShellyApiLightUtil.java:
```diff
@@ -0,0 +1,121 @@
+
+    public static Integer getLightIdFromGroup(String groupName) {
+        if (groupName.startsWith(CHANNEL_GROUP_LIGHT_INDEX)) {
+            return Integer.parseInt(substringAfter(groupName, CHANNEL_GROUP_LIGHT_INDEX)) - 1;
```
This mapping is wrong for the hybrid profiles because the `lightN` group number is not the same as the flat `settings.lights` index.

For example, in `rgbcct`:

- `settings.lights[0]` is `RGB:0` and is exposed as `control`
- `settings.lights[1]` is `CCT:0` and is exposed as `light1`

But this method maps `light1` back to index `0`. As a result, a command sent to `light1` is handled as if it targeted the RGB component instead of `CCT:0`. `rgbx2light` has the same off-by-one problem for both secondary light components.

Please make the reverse mapping profile-aware so it exactly reverses `ShellyDeviceProfile.getControlGroup()`, and add command-path tests for `rgbcct` and `rgbx2light`.

markus7017 replied:
Fixed in 9243f03: reversed `getControlGroup()` correctly by adding `profile.getColorComponentCount()` back in, so on `rgbcct`/`rgbx2light` a command to `light1`/`light2` now resolves to the right `settings.lights` index instead of the RGB component. Added regression tests for both hybrid profiles.

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2ApiRpc.java:
```diff
@@ -912,18 +942,19 @@ public void setAutoTimer(int index, String timerName, double value) throws Shell
+        int componentId = profile.isRGBW2 ? profile.getLightComponentId(index) : index;
+        Shelly2RpcRequest req = new Shelly2RpcRequest().withMethod(method).withId(componentId);
         req.params.withConfig();
         req.params.config.name = component + index;
```
The RPC target now correctly uses the component-local ID, but the generated component name still uses the flat `settings.lights` index.

For `rgbcct`, flat index `1` correctly targets `CCT:0`, but this code assigns the name `CCT1`.

Since this request only changes the auto timer, it would be safer to omit `name` from the `SetConfig` request instead of modifying an unrelated component property.

markus7017 replied:
Fixed in 9703d68363 — setAutoTimer now uses the correct config.name / CCT1 handling.

wborn reviewed (changes requested):
The latest changes address several of the previous issues. All five inputs are now included in the Pro RGBWW PM status/config processing, component-specific `ct_range` values are retained and used by the normal light-handler path, and RPC operations now translate the flat light index to the correct component-local ID.

Two functional issues still remain in the hybrid/CCT handling.

Commands to secondary `lightN` groups in `rgbcct` and `rgbx2light` are mapped back to the wrong `settings.lights` index, so they can be handled as the wrong Shelly component.

WebSocket/push CCT status updates also still use the profile-wide temperature range instead of the component-specific `ct_range`, which produces incorrect `colorTemp` percentages for customized ranges.

Requesting changes until those two remaining paths use the correct component mapping and temperature-range logic.

*This review was AI-assisted.*

andrewfg commented:
> it thought long and deep to come to this conclusion:

Crikey! I shall have to think long and deep to figure out what that means!

andrewfg commented:
I guess that we really only need to agree on the hierarchy of device/component/channel-group/channel for each product variant. My working hypothesis is shown below.

I think @markus7017 if you deliver the "architecture/framework" so the correct things are discovered and created with the correct channel-groups and channels, then I shall deliver the "logic" to produce, consume, and synchronize actual data values within those channels..

#### BULB
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, etc.
    - Color group with: r,g,b,w, full color, effect, gain, color picker etc.
    - White (cool-warm) group with: color temperature, brightness, etc

#### DUO
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, etc.
    - White (cool-warm) group with: color temperature, brightness, etc

#### VINTAGE (PS perhaps you currently have this wrong..)
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Light (brightness only) group with: explicit brightness, implicit on-off, auto-on/off 

#### RGBW2 COLOR MODE
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - Color group with: r,g,b,w, full color, effect, gain, color picker etc.
  
#### RGBW2 WHITE MODE
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- Four light components with:
    - Light (brightness only) group with: explicit brightness, implicit on-off, auto-on/off 

#### RGBWW in RGB profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - Color group with: r,g,b, ~W~ , full color, effect, gain, color picker etc.

#### RGBWW in RGBW profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - Color group with: r,g,b,w, full color, effect, gain, color picker etc.

#### RGBWW in RGBCCT profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - Color group with: r,g,b,w, full color, effect, gain, color picker etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - White (cool-warm) group with: color temperature, brightness, etc

#### RGBWW in RGBX2LIGHT profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- One light component with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - Color group with: r,g,b,w, full color, effect, gain, color picker etc.
- Two light components with:
    - Light (brightness only) group with: explicit brightness, implicit on-off, auto-on/off 

#### RGBWW in LIGHT profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- Five light components with:
    - Light (brightness only) group with: explicit brightness, implicit on-off, auto-on/off 

#### RGBWW in CCTX2 profile
- Basic device component with:
    - Status group with: firmware, signal strength, alarm, updates, last event, etc,
    - Meter group with: power, energy, etc.
- Two light components with:
    - Control group with: on-off, auto-on/off, last event, etc.
    - White (cool-warm) group with: color temperature, brightness, etc

lsiepel commented:
Shelly PR’s have the strange ability to grow and grow and have a hard time to get to the finish. 
Therefor it would be best to focus on getting support for this device done. 

That being said, the lightmodel discussion is best to move to the dedicated issue at https://github.com/openhab/openhab-addons/issues/21259

andrewfg commented:
> Therefor it would be best to focus on getting support for this device done.

@lsiepel I gave my feedback because @markus7017 asked for it. I see no point in rushing through a potentially wrong architecture and regretting it later.

andrewfg commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1740,6 +1741,71 @@ In `light` profile (white mode), each of the 4 LED output channels is exposed as
+|         | red           | Dimmer   | r/w       | Red brightness: 0..100% (control only the red channel)                  |
+|         | green         | Dimmer   | r/w       | Green brightness: 0..100% (control only the green channel)              |
+|         | blue          | Dimmer   | r/w       | Blue brightness: 0..100% (control only the blue channel)                |
+|         | white         | Dimmer   | r/w       | White brightness: 0..100% (control only the white channel)              |
```
As described here https://shelly-api-docs.shelly.cloud/gen2/Devices/Gen2/ShellyProRGBWWPM the device has 5 outputs which can be used in various combinations:

- 5x individual light -- `LIGHT` profile
- 1x RGB group, plus 2x individual light -- `RGBX2LIGHT` profile
- 1x RGB group, plus 1x cool-warm pair -- `RGBXCCT` profile
- 2x cool-warm pair -- `CCTX2` profile

Note that the first group seems always to be `RGB` and not `RGBW`.

Therefore I would assume that the RGB mode will NOT expose a White component. i.e. RGB -only..

See my comments here https://github.com/openhab/openhab-addons/pull/19227#issuecomment-5362595977

wborn replied:
There is still a concrete RGB-vs-RGBW issue here. `colorSettingsRGB` reuses the `colorFull` channel type, which still offers `white` as an option.

The Pro RGBWW PM exposes an `RGB` component, not an `RGBW` component. Selecting `white` therefore makes `handleFullColor()` produce `RGBW = 0,0,0,255`, but the RGB RPC path only sends the RGB values and ignores the white value. The device consequently receives `RGB = 0,0,0`.

An RGB-only component should therefore not expose the White option, unless White is explicitly translated to a valid RGB value.

andrewfg replied:
> the RGB RPC path only sends the RGB values and ignores the white value. The device consequently receives RGB = 0,0,0.

This is fixed by my PR

andrewfg commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1740,6 +1741,71 @@ In `light` profile (white mode), each of the 4 LED output channels is exposed as
+
+`Note`:
+`rgbcct` and `rgbx2light` combine the RGB component above with additional CCT (`rgbcct`) or Light
+(`rgbx2light`) components. Each additional component is exposed as its own `light1`/`light2` group
```
> Each additional component is exposed as its own `light1`/`light2` group

This applies only to LIGHT or RGBX2LIGHT and not to RGBCCT or CCTX2

.. also (nit) in OH markdown convention every sentence must be on a new line.

andrewfg commented on bundles/org.openhab.binding.shelly/README.md:
```diff
@@ -1740,6 +1741,71 @@ In `light` profile (white mode), each of the 4 LED output channels is exposed as
+(`rgbx2light`) components. Each additional component is exposed as its own `light1`/`light2` group
+(same layout as the `light` profile below) with its own independent meter (`meter2`/`meter3`).
+Since every profile has more than one meter, the device also gets the aggregated `device#accumulatedPower`/`device#totalEnergy` channels described in the general notes on channels above.
+
```
Probably you should add a whole new table of channels for RGBCCT with its own comments etc.

andrewfg commented on bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_lights.xml:
```diff
@@ -12,16 +12,16 @@
 		<channel-groups>
```
I think in your prior code you had used typeId `rgbwpmChannel` and now you are proposing to rename it as `rgbwChannel`.

Probably these group-types are wrong for two reasons:

1. `xyzChannel` should really be `xyzComponent`
2. It is hard to tell from the diff, but it looks like you do not have a group for all RGBWW profiles. What you need is the following:

    - RGB component (NOTE: excluding W) .. typeId `rgbComponent` 
    - CCT cool-warm component .. typeId `cctComponent`
    - LIGHT brightness-only component .. typeId `lightComponent`

See my comments in the ReadMe above.
And also here https://github.com/openhab/openhab-addons/pull/19227#issuecomment-5362595977

andrewfg commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/ShellyDevices.java:
```diff
@@ -321,7 +323,8 @@ public class ShellyDevices {
     // RGBW2 devices
     public static final Set<ThingTypeUID> GROUP_RGBW2_THING_TYPES = Set.of( //
-            THING_TYPE_SHELLYRGBW2_COLOR, THING_TYPE_SHELLYRGBW2_WHITE, THING_TYPE_SHELLYPLUSRGBWPM);
+            THING_TYPE_SHELLYRGBW2_COLOR, THING_TYPE_SHELLYRGBW2_WHITE, THING_TYPE_SHELLYPLUSRGBWPM,
```
I think RGBWWPM is not strictly an RGBW thing type since it has no W component.

See other comments above and below

andrewfg commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:
```diff
@@ -536,18 +540,14 @@ public static Map<String, Channel> createLightChannels(final Thing thing, final
         if (lights != null) {
             ShellySettingsRgbwLight light = lights.get(idx);
-            String whiteGroup = profile.isRGBW2 && !profile.inColor ? group : CHANNEL_GROUP_WHITE_CONTROL;
+            String whiteGroup = profile.isRGBW2 && !profile.hasColorTag(idx) ? group : CHANNEL_GROUP_WHITE_CONTROL;
```
See comments about difference between RGB and RGBW

andrewfg commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/provider/ShellyChannelDefinitions.java:
```diff
@@ -246,6 +246,7 @@ public ShellyChannelDefinitions(@Reference ShellyTranslationProvider translation
                 .add(new ShellyChannel(m, CHGR_LIGHTCH, CHANNEL_TIMER_ACTIVE, "timerActive", ITEMT_SWITCH))
                 // RGBW2-white / RGBW PM-white
                 .add(new ShellyChannel(m, CHGR_LIGHT_IDX, CHANNEL_BRIGHTNESS, "whiteBrightness", ITEMT_DIMMER))
+                .add(new ShellyChannel(m, CHGR_LIGHT_IDX, CHANNEL_COLOR_TEMP, "whiteTemp", ITEMT_DIMMER))
```
I think not. I think it is as follows..

- LIGHT components have a `LIGHT` group which never has a color temperature channel. It only has explicit brightness (and implicit on-off).
- CCT components have `WHITE` group which has brightness, cool-warm color temperature, .. and perhaps either a) a `CONTROL` group, or b) includes implicit on-off, and auto-on/off etc.

andrewfg commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyLightHandler.java:
Don't worry too much about the logic here. My PR almost certainly re-writes it..

andrewfg reviewed (changes requested):
I have a few comments, summarised as follows:
1. in RGBWW PM if the first component is a color component it is RGB:n and not RGBW:n i.e. no white channel.
2. in RGBWW PM any CCT:n component is a `WHITE` group with cool-warm, and brightness (plus either implicit, or explicit on-off, and auto-on/off, etc.)
3. in RGBWW PM any LIGHT:n component is a `LIGHT` group with brightness (plus either implicit, or explicit on-off, and auto-on/off, etc.)

See this https://shelly-api-docs.shelly.cloud/gen2/Devices/Gen2/ShellyProRGBWWPM

markus7017 commented:
I agree with @lsiepel, we should focus on fixes here, create a new pr to complete the basics and then integrate light control and finally the Duo bulb

markus7017 commented:
@wborn Pushed fixes for the two remaining issues from your last review:

- **Fix:** `ShellyApiLightUtil.getLightIdFromGroup()` now reverses `getControlGroup()` correctly — it adds `profile.getColorComponentCount()` back in, so on `rgbcct`/`rgbx2light` a command sent to `light1`/`light2` resolves to the right `settings.lights` index instead of being off by the leading color component. Replied inline.
- **Fix:** `updateLightMode()`'s WebSocket/push-status path now converts `light.temp` with the per-component `profile.getMinTemp(i)`/`getMaxTemp(i)` instead of the profile-wide default, matching the polling path.

Also found and fixed a related issue while in there: `Shelly2ApiRpc.setAutoTimer()` was building `config.name` from the flat `settings.lights` index, which doesn't match the on-device component id on hybrid profiles — dropped that field entirely since this call only ever changes the auto-timer, and removed the now-unused `displayName` field it depended on.

Full build (spotless, tests, SAT) passes; added regression tests for both hybrid profiles.

Re: the channel/group modeling discussion — agreed with @lsiepel above to keep this PR scoped to fixes and move that to a follow-up PR.

andrewfg commented:
> move that to a follow-up PR.

@markus7017 please do not open another PR yourself. Instead I will include any necessary changes or adaptions in my own PR. Otherwise we have too many independently moving parts.

markus7017 commented:
this one is already big, I think we should have a step in between, but up to you

Anything else for this one?

andrewfg commented:
> Anything else for this one?

No.

> think we should have a step in between, but up to you

I am trying to parallel track most of what you are doing here and in the Gen 3 light PR, but juggling a 3 way merge is hard; and a 4 way merge would be harder yet. So let's finish this and the Gen 3 lights asap, and then I have a stable base to work from.

markus7017 commented:
uff, the is a multi dimensional feature matrix 🤪
will fiy this later

markus7017 commented:
> So let's finish this and the Gen 3 lights asap, and then I have a stable base to work from.

ok, let's merge this one, then I'll rebase & review the  Duo PR and you take it fron there

andrewfg commented:
> will fiy this later (assuming you meant "fix" ..)

@markus7017 This error here is about removing the excess W channel from the group. My PR is (so far) focussed on the channel command/update process assuming that the actual list of groups and channels were correct from your side.

Therefore I think we could split this removal process into two parts:

1. remove the excess W channel from thing-type xml and the `createChannels()` method
2. my PR ignores any W channel in this case (it already ignores it)

So part 2. is already solved in my scope. The only question is if a) you want to remove the excess W channel from thing-type xml and from `createChannels()` in this PR?, or b) if I shall fix the thing-type.xml and `createChannels()` method in my PR? (As mentioned before, I think creating a third piggy- in-the-middle PR makes no sense). I would slightly prefer a) if possible, but the alternative is if you just give me the code snippets to include in mine. As you wish.

markus7017 commented:
Fixed the excess white channel: Pro RGBWW's `rgb:0` has no white output, so `color` now uses a dedicated `colorSettingsRGB` group-type (no `white` channel) instead of reusing RGBW's. Went with option (a), scoped to this PR. Also replied on the two open threads — both already fixed on the branch (d33981ed13, 9703d68363).

wborn commented on bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/handler/ShellyLightHandler.java:
```diff
@@ -130,7 +131,7 @@ public boolean handleDeviceCommand(ChannelUID channelUID, Command command) throw
                     col.setGain(setColor(lightId, SHELLY_COLOR_GAIN, command, SHELLY_MIN_GAIN, SHELLY_MAX_GAIN));
```
The command is sent to the correct secondary CCT/Light component, but the optimistic state update afterward targets the wrong openHAB channel.

For example, in the `rgbcct` profile:

`control` = RGB component  
`light1` = CCT component

A command such as `light1#brightness = 50` is correctly sent to the CCT component, but this code then updates `control#power` to ON. That incorrectly changes the reported state of the independent RGB component.

The ON/OFF brightness path has a similar issue because it updates the hard-coded `white#brightness` group instead of the current `lightN` group.

Please derive the optimistic update target from the current component/group and add a regression test for brightness commands on a secondary component of a hybrid profile.

andrewfg replied:
@markus7017 I don't really know how to help you here; but I can tell you that this ongoing AI review is forcing you to write convoluted code that will be immediately replaced in my PR with the LightModel. Indeed a lot of the code that you have already been forced to write has already been directly replaced by my LightModel code. In other words I have the platform to immediately build on this PR and solve all the open issues. But unfortunately I cannot do that while we are on parallel branch/forks with merge conflicts that require manual merging. And we already learned that a three way git merge between OH main, your fork/branch and my fork/branch is way beyond my skill set.. So I would like to help you, but I cannot. => Perhaps @lsiepel or @wborn can make a suggestion how to proceed?

wborn replied:
Since #21286 currently depends on #19227 and replaces much of this light handling, the AI suggests keeping this fix narrow rather than introducing any further component-model changes here.

For the ON/OFF path, the existing helper already seems to provide the correct group for both the legacy devices and the new per-component groups:

```diff
- updateChannel(CHANNEL_COLOR_WHITE, CHANNEL_BRIGHTNESS + "$Switch", col.power);
- updateChannel(CHANNEL_COLOR_WHITE, CHANNEL_BRIGHTNESS + "$Value", toQuantityType(
+ String brightnessGroup = buildWhiteGroupName(profile, lightId);
+ updateChannel(brightnessGroup, CHANNEL_BRIGHTNESS + "$Switch", col.power);
+ updateChannel(brightnessGroup, CHANNEL_BRIGHTNESS + "$Value", toQuantityType(
        (double) (col.power == OnOffType.ON ? col.brightness : 0), DIGITS_NONE, Units.PERCENT));
```

For numeric brightness commands, it seems sufficient to avoid the legacy optimistic `control#power` update when the command belongs to a Pro RGBWW PM secondary component:

```diff
- updateChannel(CHANNEL_GROUP_LIGHT_CONTROL, CHANNEL_LIGHT_POWER, OnOffType.from(value > 0));
+ if (!profile.isProRgbwwPm || CHANNEL_GROUP_LIGHT_CONTROL.equals(groupName)) {
+     updateChannel(CHANNEL_GROUP_LIGHT_CONTROL, CHANNEL_LIGHT_POWER, OnOffType.from(value > 0));
+ }
```

This keeps the existing behavior for other devices while preventing a secondary `lightN` command from changing the reported state of the independent RGB `control` component.

WDYT?

markus7017 replied:
Applied — commit be7d02421f. Both changes as suggested: the switch-brightness path now targets `buildWhiteGroupName(profile, lightId)` instead of the hard-coded `control` white group, and the numeric-brightness path only touches `control#power` when the command is already targeting the control group, so a secondary `lightN` command on a hybrid profile no longer clobbers the independent RGB component's reported state.

andrewfg commented:
> In particular, the RGB-only color group still exposes the shared colorFull White option, although that value cannot be represented by the device's RGB component.

That AI statement is wrong!! RGB 255,255,255 is white!!

markus7017 commented:
On the colorFull white debate: confirmed against the RGB.Set/GetStatus API spec — RGB component has only `rgb:[r,g,b]` + `brightness`, no white field. Root bug: colorFull's "white" set `rgb=0,0,0` (white value silently dropped for the RGB-only component), turning the light off. Fixed in a928c4ac: "white" now mixes `rgb:[255,255,255]` for this component, and status reporting recognizes that combination as "white" too.

lsiepel commented:
I tend to merge this as is. It is another good step forward with good quality. Yes, there might be a corner case left, we should treat the PR from @andrewfg as the one that addresses andy leftovers. 
Without any feedback, i'll move forward this evening.

markus7017 commented:
@lsiepel 
yes, please go ahead

lsiepel commented:
> @lsiepel yes, please go ahead

Well the change suggested by@wborn here https://github.com/openhab/openhab-addons/pull/19227#discussion_r3838665952 is a small one that I recommend to fix here. The rest will be handled in the core lighting pr

andrewfg commented:
@markus7017 I am seeing "Unknown API value 'rgbcct'" during my testing. Admittedly during testing my own PR (where maybe I broke something). But just wondering if you see it too?

```
2026-08-25 11:27:50.561 [TRACE] [shelly.internal.api.ShellyHttpClient] - shellycolorblbg3-90706944d8ec: HTTP GET http://192.168.1.161/shelly
Accept-Encoding: gzip
User-Agent: Jetty/9.4.58.v20250814
Content-Type: application/x-www-form-urlencoded



2026-08-25 11:27:50.577 [TRACE] [shelly.internal.api.ShellyHttpClient] - shellyprorgbwwpm-0070074b2c88: HTTP Response 200: {"name":null,"id":"shellyprorgbwwpm-0070074b2c88","mac":"0070074B2C88","slot":1,"model":"SPDC-0D5PE16EU","gen":2,"fw_id":"20241021-105029/1.4.99-prorgbwwpmprod2-g115f3e9","ver":"1.4.99-prorgbwwpmprod2","app":"ProRGBWWPM","auth_en":false,"auth_domain":null,"profile":"rgbcct"}
Content-Type: application/json
Content-Length: 275
Pragma: no-cache
Server: ShellyHTTP/1.0.0
Connection: close


2026-08-25 11:27:50.578 [WARN ] [helly.internal.api2.Shelly2ApiClient] - shellyprorgbwwpm-0070074b2c88: Unknown API value 'rgbcct' (map data={triphase=emeter, light=white, switch=relay, monophase=emeter, rgbw=color, cover=roller, rgb=color}), please create an issue on GitHub
```
